### DISCUSSION TITLE
Note completion feature, like Github Copilot for Code

### DISCUSSION THREAD

**Discussion by artofbeinghuman** [brianpetro/obsidian-smart-connections#459] [Ideas]
Hi,

I use obsidian to write in my diary. I always wondered if an AI couldn't help me in writing down my thoughts if it where finetuned or had RAG-access to all the diary entries of the past years.

What about an "inline" feature of smart connections plugin that uses the local context of what I wrote so far and knowledge about all my other notes to suggest a completion of my current thoughts.

I'm basically thinking of what we have with Github Copilot for coding, but the same for my own individual thoughts as condition via my past diary entries.

Cheers,
Marvin

**Comment by micuentadecasa** [comment]
I vote for this.

**Comment by brianpetro** [comment]
> I'm basically thinking of what we have with Github Copilot for coding, but the same for my own individual thoughts as condition via my past diary entries.

I find myself thinking about this a lot.

GitHub copilot uses a different type of generative model, "instruct" rather than "chat," so the current embeddings/retrieval doesn't exactly transfer to the "predict next" type of implementation we would want here.

But this is something I too would love to have in Obsidian. So, if I think I can make it work, I will.

One of the main challenges is the requirement of a significantly larger embedding space. I won't get into the details, but since notes don't have the same well-formed structure as code, it requires embedding the same content not just once but many overlapping times, with more overlapping embeddings improving the results. Stronger models may reduce this requirement, but that would significantly increase cost because of the context requirement for every prediction.

🌴

**Comment by Richie-Peak** [comment]
I am strongly against this assuming it would take the amount of dev work i think it would. 

The specific idea of using an AI model to attempt to complete diary entries for you makes absolutely no sense: the entire point of a diary is to express YOUR unique feelings. It's for you to verbalise your thoughts and feelings. I am an AI can do that without destroying the key advantages of journaling. By definition it wouldn't be helping your get your thoughts out, it would be suggesting that you lazily cram your thoughts into a predefined form. 

Note completion is not at all like code completion. In code completion, an expert can predict what code you will write as you're writing. When you write a journal, no one can do that.

**Reply by brianpetro** [reply]
@Richie-Peak 

Perhaps I should've clarified that my interpretation of "journalling," as mentioned by the OP, was just the feature that could be used on notes in general. 

In that case, many of my notes are for project management. These project notes tend to evolve similarly throughout their lifecycle to completion, and that's where I see something like this being beneficial for me.

I hope that makes sense. 

But yes, I agree that the same feature could be detrimental to the desired benefits of journalling.

🌴

**Comment by andrexibiza** [comment]
<img width="599" alt="image" src="https://github.com/brianpetro/obsidian-smart-connections/assets/84248988/f923861b-288b-4147-af00-5d8bb260b235">

The "Companion" extension already offers this autocomplete feature. Personally, I found it to be extremely annoying, but it definitely does work.

**Reply by brianpetro** [reply]
It's cool that they figured out how to implement the copilot-esque UI/UX with the faded text.

I've been playing around with a prediction feature in `v2.1` that uses similar notes as context to better predict what you might put in the note. It's also available a command, but the UI/UX is more similar to text generator than copilot. 

🌴
